File: codes/2023/25-WildcardExcellence-Beihang (Java)/src/midend/Type.java
package midend;

/*
IR的类型系统，包括：
一切都是Value，每个Value会有一个Type
VOID：无返回值
INTEGER：整数 i32, i1
<<<<<<< HEAD
<<<<<<< HEAD
FloatType：浮点数 float
=======
>>>>>>> d499c19 (基本中端数据结构)
=======
FloatType：浮点数 float
>>>>>>> 3869b3e (浮点数)
POINTER：指针
ARRAY：数组
 */

public class Type {

    public int getLength() {
        return 1;
    }

    public int getSize() {
        return 1;
    }

    public Type getElementType() {
        return null;
    }

    public boolean isIntegerType() {
        return this instanceof IntegerType;
    }

    public boolean isPointerType() {
        return this instanceof PointerType;
    }

    public boolean isVoidType() {
        return this instanceof VoidType;
    }

    public boolean isArrayType() {
        return this instanceof ArrayType;
    }

    public boolean isFloatType() {
        return this instanceof FloatType;
    }

    public boolean isBasicType() {
        return !(isArrayType() || isPointerType());
    }

    public static class PointerType extends Type {
        private final Type elementType;

        public PointerType(Type elementType) {
            this.elementType = elementType;
        }

        public Type getElementType() {
            return this.elementType;
        }

        public String toString() {
            return elementType.toString() + "*";
        }
    }

    public static class IntegerType extends Type {
        private final String name; // i32 / i1
        public static final IntegerType I32 = new IntegerType("i32");
        public static final IntegerType I1 = new IntegerType("i1");

        public IntegerType(String name) {
            this.name = name;
        }

        public boolean isI1() {
            return this.name.equals("i1");
        }

        public boolean isI32() {
            return this.name.equals("i32");
        }

        public String toString() {
            return this.name;
        }
    }

    public static class FloatType extends Type {
        public FloatType() {

        }

        public String toString() {
            return "float";
        }
    }

    public static class ArrayType extends Type {
        private final Type elementType;
        private final int size;

        public ArrayType(Type elementType, int size) {
            this.elementType = elementType;
            this.size = size;
        }

        public Type getElementType() {
            return this.elementType;
        }

        @Override
        public int getLength() {
            return this.size * elementType.getLength();
        }

        public int getSize() {
            return this.size;
        }

        public String toString() {
            return "[" + size + " x " + elementType.toString() + "]";
        }
    }

    public static class VoidType extends Type {
        public VoidType() {
        }

        public String toString() {
            return "void";
        }
    }
}

File: codes/2025/7-Second-BUPT-100-56-65 (C++)/src/IR/Type.cpp
#include "IR/Type.h"

#include <iostream>
#include <sstream>

#include "IR/Value.h"

namespace midend {

IntegerType* IntegerType::get(Context* ctx, unsigned bits) {
    return ctx->getIntegerType(bits);
}

FloatType* FloatType::get(Context* ctx) { return ctx->getFloatType(); }

VoidType* VoidType::get(Context* ctx) { return ctx->getVoidType(); }

LabelType* LabelType::get(Context* ctx) { return ctx->getLabelType(); }

PointerType* PointerType::get(Type* elemType) {
    return elemType->getContext()->getPointerType(elemType);
}

ArrayType* ArrayType::get(Type* elemType, uint64_t numElems) {
    return elemType->getContext()->getArrayType(elemType, numElems);
}

FunctionType* FunctionType::get(Type* retType, std::vector<Type*> params,
                                bool varArg) {
    return new FunctionType(retType, std::move(params), varArg,
                            retType->getContext());
}

std::string FunctionType::toString() const {
    std::stringstream ss;
    ss << returnType_->toString() << " (";
    for (size_t i = 0; i < paramTypes_.size(); ++i) {
        if (i > 0) ss << ", ";
        ss << paramTypes_[i]->toString();
    }
    if (isVarArg_) {
        if (!paramTypes_.empty()) ss << ", ";
        ss << "...";
    }
    ss << ")";
    return ss.str();
}

Context::Context() {
    voidType_.reset(new VoidType(this));
    floatType_.reset(new FloatType(this));
    labelType_.reset(new LabelType(this));
}

VoidType* Context::getVoidType() { return voidType_.get(); }

IntegerType* Context::getIntegerType(unsigned bits) {
    if (bits != 1 && bits != 32 && bits != 8) {
        std::cerr << "Warning: unsupported integer type bits: " << bits
                  << ". Only 1, 8, and 32 bits are supported." << std::endl;
    }
    auto& type = integerTypes_[bits];
    if (!type) {
        type.reset(new IntegerType(bits, this));
    }
    return type.get();
}

FloatType* Context::getFloatType() { return floatType_.get(); }

LabelType* Context::getLabelType() { return labelType_.get(); }

PointerType* Context::getPointerType(Type* elemType) {
    auto& type = pointerTypes_[elemType];
    if (!type) {
        type.reset(new PointerType(elemType, this));
    }
    return type.get();
}

ArrayType* Context::getArrayType(Type* elemType, uint64_t numElems) {
    std::string key =
        elemType->toString() + "[" + std::to_string(numElems) + "]";
    auto& type = arrayTypes_[key];
    if (!type) {
        type.reset(new ArrayType(elemType, numElems, this));
    }
    return type.get();
}

PointerType* IntegerType::getPointerTo() { return PointerType::get(this); }

PointerType* FloatType::getPointerTo() { return PointerType::get(this); }

PointerType* VoidType::getPointerTo() { return PointerType::get(this); }

Type* Type::getSingleElementType() const {
    if (isArrayType()) {
        auto* arrayType = static_cast<const ArrayType*>(this);
        return arrayType->getElementType();
    }
    if (isPointerType()) {
        auto* ptrType = static_cast<const PointerType*>(this);
        return ptrType->getElementType();
    }
    return const_cast<Type*>(this);
}

Type* Type::getBaseElementType() const {
    if (isArrayType()) {
        auto* arrayType = static_cast<const ArrayType*>(this);
        return arrayType->getElementType()->getBaseElementType();
    }
    if (isPointerType()) {
        auto* ptrType = static_cast<const PointerType*>(this);
        return ptrType->getElementType()->getBaseElementType();
    }
    return const_cast<Type*>(this);
}

Type* Type::getMultiLevelElementType(unsigned levels) const {
    Type* current = const_cast<Type*>(this);
    for (unsigned i = 0; i < levels; ++i) {
        current = current->getSingleElementType();
    }
    return current;
}

unsigned Type::getSizeInBytes() const {
    if (isIntegerType()) {
        auto* intTy = static_cast<const IntegerType*>(this);
        return (intTy->getBitWidth() + 7) / 8;
    }
    if (isFloatType()) {
        return 4;  // float = 4 bytes
    }
    if (isPointerType()) {
        return 8;  // pointer = 8 bytes (64-bit)
    }
    if (isArrayType()) {
        auto* arrayTy = static_cast<const ArrayType*>(this);
        return arrayTy->getNumElements() *
               arrayTy->getElementType()->getSizeInBytes();
    }
    // For other types (void, label, etc.), return 0
    return 0;
}

}  // namespace midend